code/go/0chain.net/smartcontract/storagesc/challenge.go

```diff
@@ -722,39 +827,77 @@ func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
+	if err != nil {
+		if err == util.ErrValueNotPresent {
+			blobberChallengeObj = &BlobberChallenge{}
+			blobberChallengeObj.BlobberID = blobberID
+		} else {
+			return "", common.NewError("add_challenge",
+				"error fetching blobber challenge: "+err.Error())
+		}
+	}
 
-	blobberChallengeObj := &BlobberChallenge{}
-	blobberChallengeObj.BlobberID = storageChallenge.Blobber.ID
+	addedChallenge := blobberChallengeObj.addChallenge(&storageChallenge)
+	if !addedChallenge {
+		challengeBytes, err := json.Marshal(storageChallenge)
+		return string(challengeBytes), err
+	}
 
-	blobberChallengeBytes, _ := balances.GetTrieNode(blobberChallengeObj.GetKey(sc.ID))
-	if blobberChallengeBytes != nil {
-		err = blobberChallengeObj.Decode(blobberChallengeBytes.Encode())
-		if err != nil {
-			return "", common.NewError("blobber_challenge_decode_error",
-				"Error decoding the blobber challenge")
+	allocChallengeObj, err := sc.getAllocationChallenge(alloc.ID, balances)
+	if err != nil {
+		if err == util.ErrValueNotPresent {
+			allocChallengeObj = &AllocationChallenge{}
+			allocChallengeObj.AllocationID = alloc.ID
+		} else {
+			return "", common.NewError("add_challenge",
+				"error fetching allocation challenge: "+err.Error())
 		}
 	}
 
-	storageChallenge.Created = creationDate
-	addedChallege := blobberChallengeObj.addChallenge(&storageChallenge)
-	if !addedChallege {
+	addedAllocChallenge := allocChallengeObj.addChallenge(&storageChallenge)
+	if !addedAllocChallenge {
 		challengeBytes, err := json.Marshal(storageChallenge)
 		return string(challengeBytes), err
 	}
 
-	balances.InsertTrieNode(blobberChallengeObj.GetKey(sc.ID), blobberChallengeObj)
+	_, err = balances.InsertTrieNode(allocChallengeObj.GetKey(sc.ID), allocChallengeObj)
+	if err != nil {
+		return "", common.NewError("add_challenge",
+			"error storing alloc challenge: "+err.Error())
+	}
+
+	_, err = balances.InsertTrieNode(storageChallenge.GetKey(sc.ID), &storageChallenge)
+	if err != nil {
+		return "", common.NewError("add_challenge",
+			"error storing challenge: "+err.Error())
+	}
+
+	_, err = balances.InsertTrieNode(blobberChallengeObj.GetKey(sc.ID), blobberChallengeObj)
+	if err != nil {
+		return "", common.NewError("add_challenge",
+			"error storing blobber challenge: "+err.Error())
+	}
 
 	alloc.Stats.OpenChallenges++
 	alloc.Stats.TotalChallenges++
 	blobberAllocation.Stats.OpenChallenges++
 	blobberAllocation.Stats.TotalChallenges++
-	balances.InsertTrieNode(alloc.GetKey(sc.ID), alloc)
+
+	_, err = balances.InsertTrieNode(alloc.GetKey(sc.ID), alloc)
+	if err != nil {
+		return "", common.NewError("add_challenge",
+			"error storing allocation: "+err.Error())
+	}
 	//Logger.Info("Adding a new challenge", zap.Any("blobberChallengeObj", blobberChallengeObj), zap.Any("challenge", storageChallenge.ID))
 	challengeBytes, err := json.Marshal(storageChallenge)
 	sc.newChallenge(balances, storageChallenge.Created)
```

Comment:
`sc.newChallenge(balances, storageChallenge.Created)` I am not convinced we need this, but it does not need to be updated once per loop, we can do this once at the end outside the for loops.